Add compression level support to Zip Module

mashumafi opened the pull request:
<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

I am working on a project that requires a file to be uncompressed.
The project is a epub tool which requires a file called `mimetype` be uncompressed as part of the specification.

Added some tests to verify the size is indeed different based on the compression level used. Also tested the reader to prevent regressions even though this PR does not touch it.

mashumafi committed:
Add compression level support to ZipPacker

Calinou commented on modules/zip/zip_packer.h:
```diff
@@ -52,9 +53,18 @@ class ZIPPacker : public RefCounted {
 		APPEND_ADDINZIP = 2,
 	};
 
+	enum CompressionLevel {
+		COMPRESSION_DEFAULT = 0,
+		COMPRESSION_NONE = 1,
+		COMPRESSION_BEST_SPEED = 2,
+		COMPRESSION_BEST = 3,
+	};
```
If values are ordered, the numbers can be removed:

```suggestion
	enum CompressionLevel {
		COMPRESSION_DEFAULT,
		COMPRESSION_NONE,
		COMPRESSION_BEST_SPEED,
		COMPRESSION_BEST,
	};
```

Calinou commented on modules/zip/zip_packer.cpp:
```diff
@@ -115,9 +133,16 @@ void ZIPPacker::_bind_methods() {
 	BIND_ENUM_CONSTANT(APPEND_CREATEAFTER);
 	BIND_ENUM_CONSTANT(APPEND_ADDINZIP);
+
+	BIND_ENUM_CONSTANT(COMPRESSION_NONE);
+	BIND_ENUM_CONSTANT(COMPRESSION_DEFAULT);
+	BIND_ENUM_CONSTANT(COMPRESSION_BEST);
+	BIND_ENUM_CONSTANT(COMPRESSION_BEST_SPEED);
```
List bindings in the same order as the enum, so that the order in the class reference XML is consistent:

```suggestion
	BIND_ENUM_CONSTANT(COMPRESSION_DEFAULT);
	BIND_ENUM_CONSTANT(COMPRESSION_NONE);
	BIND_ENUM_CONSTANT(COMPRESSION_BEST_SPEED);
	BIND_ENUM_CONSTANT(COMPRESSION_BEST);
```

Calinou commented on modules/zip/doc_classes/ZIPPacker.xml:
```diff
@@ -72,5 +79,17 @@
 			Add new files to the existing zip archive at the given path.
 		</constant>
+		<constant name="COMPRESSION_NONE" value="1" enum="CompressionLevel">
+			Start a file with no compression.
```
```suggestion
			Start a file with no compression (also known as the "Store" compression mode).
```

Calinou commented on modules/zip/doc_classes/ZIPPacker.xml:
```diff
@@ -72,5 +79,17 @@
+			Start a file with no compression.
+		</constant>
+		<constant name="COMPRESSION_DEFAULT" value="0" enum="CompressionLevel">
+			Start a file with the default compression level.
```
```suggestion
			Start a file with the default Deflate compression level ([code]6[/code]). This is a good compromise between speed and file size.
```

Calinou commented on modules/zip/doc_classes/ZIPPacker.xml:
```diff
@@ -72,5 +79,17 @@
+			Start a file with the default compression level.
+		</constant>
+		<constant name="COMPRESSION_BEST" value="3" enum="CompressionLevel">
+			Start a file with the the best compression level.
```
```suggestion
			Start a file with the the best Deflate compression level ([code]9[/code]). This is slow to compress, but results in smaller file sizes. Decompression speed is mostly unaffected.
```

Calinou reviewed (commented):
Tested locally, it works as expected.

**Testing project:** [test_zip_compression_level.zip](https://github.com/user-attachments/files/19061386/test_zip_compression_level.zip)

Compressing a single 10.5 MB text file (1 million times the string "Hello world!") with each compression mode. From top to bottom: Default, None, Best Speed (Fast), Best:

![image](https://github.com/user-attachments/assets/bbe51ca0-d2d2-4f87-9a89-7a67ca8f387c)

Most of the time, Best will not be any better than Default, but it's significantly slower to compress, so it should be used with care.

Some feedback:

- I suggest adding a getter for `compression_level` and exposing it as a property, so you can write `ZIPPacker.compression_level = ...`.
- I suggest renaming `COMPRESSION_BEST_SPEED` to `COMPRESSION_FAST`, as "fast" is generally understood to be the counterpart of "best" when it comes to compression algorithms. It also avoids confusion with the existing `COMPRESSION_BEST`.

mashumafi commented:
> I suggest adding a getter for compression_level and exposing it as a property, so you can write ZIPPacker.compression_level = ....

@Calinou 

Would it be worth exposing compression_level as a `int` instead of `enum`. I think having the `enum` is nice so users have. some reference but maybe they might want to use a direct compression number instead?

Calinou commented:
> Would it be worth exposing compression_level as a int instead of enum. I think having the enum is nice so users have. some reference but maybe they might want to use a direct compression number as well? We can avoid the switch statement which would be nice.

That makes sense, yes. We should probably keep some constants around for the default value (`-1`) and no compression (`0`) still.

I originally had the concern we might want to change the underlying compression algorithm. However, since this is ZIP and DEFLATE is the only universally supported compression method in ZIP, it's fair to assume it won't change anytime soon.

mashumafi commented:
Thanks for the feedback! 3 changes since last review:
* Renamed `COMPRESSION_BEST_SPEED` to `COMPRESSION_FAST`
* `compression_level` property, which is an `int`, `enum CompressionLevel` still exists for reference
* `ZIPReader::get_compression_level` exposed with a test to confirm it returns correctly, which works similar to `file_exists` but strictly meant to determine the compression level used. Could be useful for validation of compression levels if someone needs it.

Calinou commented:
**Updated testing project:** [test_zip_compression_level_2.zip](https://github.com/user-attachments/files/19170743/test_zip_compression_level_2.zip)

File sizes for each compression level:

```
11,001,796 archive0.zip
    53,473 archive1.zip
    53,473 archive2.zip
    53,473 archive3.zip
    21,494 archive4.zip
    21,494 archive5.zip
    21,494 archive7.zip
    21,494 archive8.zip
    21,494 archive9.zip
```

Using a compression level outside the allowed range (e.g. `-2` or `10`) will result in a 61-byte ZIP archive that can't be opened. This should print an error instead.

PS: Interestingly, using the `file` command on archives created with all compression levels will always says that the compression method is Deflate, even for the one that only contains files with level 0 compression (Store):

```
archive-2.zip: Zip archive data, at least v2.0 to extract, compression method=deflate
archive0.zip:  Zip archive data, at least v2.0 to extract, compression method=deflate
archive-1.zip: Zip archive data, at least v2.0 to extract, compression method=deflate
archive1.zip:  Zip archive data, at least v2.0 to extract, compression method=deflate
archive2.zip:  Zip archive data, at least v2.0 to extract, compression method=deflate
archive3.zip:  Zip archive data, at least v2.0 to extract, compression method=deflate
archive4.zip:  Zip archive data, at least v2.0 to extract, compression method=deflate
archive5.zip:  Zip archive data, at least v2.0 to extract, compression method=deflate
archive7.zip:  Zip archive data, at least v2.0 to extract, compression method=deflate
archive8.zip:  Zip archive data, at least v2.0 to extract, compression method=deflate
archive9.zip:  Zip archive data, at least v2.0 to extract, compression method=deflate
archive10.zip: Zip archive data, at least v2.0 to extract, compression method=deflate
```

mashumafi commented:
Added a check that `compression_level` can only be between `-2` and `10`

Calinou commented on modules/zip/zip_packer.cpp:
```diff
@@ -55,6 +55,15 @@ Error ZIPPacker::close() {
 }
 
+void ZIPPacker::set_compression_level(int p_compression_level) {
+	ERR_FAIL_COND_MSG(Z_DEFAULT_COMPRESSION > p_compression_level || p_compression_level > Z_BEST_COMPRESSION, "Invalid compression level.");
```
While I understand the indent behind the [Yoda condition](https://en.wikipedia.org/wiki/Yoda_conditions) here, we generally avoid this form in Godot's codebase, so this is preferable:

```suggestion
	ERR_FAIL_COND_MSG(p_compression_level < Z_DEFAULT_COMPRESSION || p_compression_level > Z_BEST_COMPRESSION, "Invalid compression level.");
```

AThousandShips commented on modules/zip/doc_classes/ZIPPacker.xml:
```diff
@@ -72,5 +77,17 @@
+			Start a file with the default Deflate compression level ([code]6[/code]). This is a good compromise between speed and file size.
+		</constant>
+		<constant name="COMPRESSION_NONE" value="0" enum="CompressionLevel">
+			Start a file with no compression. This is also known as the "Store" compression mode and is the fastest method of packing files inside a ZIP archive. Consider using this mode for files that are already compressed (such as JPEG, PNG, MP3 or Ogg Vorbis files).
```
```suggestion
			Start a file with no compression. This is also known as the "Store" compression mode and is the fastest method of packing files inside a ZIP archive. Consider using this mode for files that are already compressed (such as JPEG, PNG, MP3, or Ogg Vorbis files).
```
Please use [Oxford comma](https://docs.godotengine.org/en/latest/contributing/documentation/docs_writing_guidelines.html#use-the-oxford-comma-to-enumerate-anything).

AThousandShips commented on modules/zip/doc_classes/ZIPReader.xml:
```diff
@@ -61,6 +61,16 @@
+			<param index="0" name="path" type="String" />
+			<param index="1" name="case_sensitive" type="bool" default="true" />
+			<description>
+				Returns the compression level of the file in the loaded zip archive.
+				Returns [code]-1[/code] if the file doesn't exist or any other error occurs.
+				Must be called after [method open].
```
```suggestion
				Returns the compression level of the file in the loaded zip archive. Returns [code]-1[/code] if the file doesn't exist or any other error occurs. Must be called after [method open].
```
Each line is a new paragraph, think this is better this way

mashumafi replied:
Instead of 1 line I chose 2 to be consistent with the rest of the docs. I can update the other docs though if you prefer it all in one paragraph.

mashumafi replied:
updated to 1 line

Repiteo reviewed (approved):
Looks good!

Repiteo commented:
Thanks!
